Metric spaces (#1162)
UniMath/agda-unimath: src/metric-spaces/precategory-of-metric-spaces-and-isometries.lagda.md

  precategory-isometry-Metric-Space =
    make-Precategory
      ( Metric-Space l1 l2)
      ( set-isometry-Metric-Space)
      ( λ {A B C} → comp-isometry-Metric-Space A B C)
      ( isometry-id-Metric-Space)
      ( λ {A B C D} h g f →
        eq-htpy-map-isometry-Metric-Space
          ( A)
          ( D)
          ( comp-isometry-Metric-Space A B D
            ( comp-isometry-Metric-Space B C D h g)
            ( f))
          ( comp-isometry-Metric-Space A C D
            ( h)
            ( comp-isometry-Metric-Space A B C g f))
          ( λ x → refl))
      ( λ {A B} f →
        eq-htpy-map-isometry-Metric-Space A B
          ( comp-isometry-Metric-Space A B B
            ( isometry-id-Metric-Space B)
            ( f))
          ( f)
          ( λ x → refl))
      ( λ {A B} f →
        eq-htpy-map-isometry-Metric-Space A B
          ( comp-isometry-Metric-Space A A B
            ( f)
            ( isometry-id-Metric-Space A))
          ( f)
          ( λ x → refl))
```

## Properties

### Isomorphisms in the precategory of metric spaces and isometries are equivalences

```agda
module _
  {l1 l2 : Level} (A B : Metric-Space l1 l2)
  (f : isometry-Metric-Space A B)
  where

>   is-equiv-is-iso-precatory-isometry-Metric-Space :

Reviewer: ```suggestion
  is-equiv-is-iso-isometry-Metric-Space :
```

> module _
>   {l1 l2 : Level} (A B : Metric-Space l1 l2)
>   (f : isometry-Metric-Space A B)
>   where
> 
>   is-equiv-is-iso-precatory-isometry-Metric-Space :
>     is-iso-Precategory precategory-isometry-Metric-Space {A} {B} f →
>     is-equiv (map-isometry-Metric-Space A B f)

Reviewer: Building on my comment about the category of sets, here again there's an issue with universe polymorphism, although large categories does not resolve it directly. Namely, your lemma doesn't apply to metric spaces of differing universe levels. The best solution for now is probably to replace `is-iso-Precategory` by the type written out as follows

```suggestion
module _
  {l1 l2 l3 l4 : Level} (A : Metric-Space l1 l2) (B : Metric-Space l3 l4)
  (f : isometry-Metric-Space A B)
  where

  is-equiv-is-iso-isometry-Metric-Space :
    Σ ( isometry-Metric-Space B A)
      ( λ g →
        ( comp-isometry-Metric-Space {B} {A} {B} f g ＝
          id-isometry-Metric-Space B) ×
        ( comp-isometry-Metric-Space {A} {B} {A} g f ＝
          id-isometry-Metric-Space B) )
    is-iso-Precategory precategory-isometry-Metric-Space {A} {B} f →
    is-equiv (map-isometry-Metric-Space A B f)
```

note also the name change

Reviewer: On second thought, nevermind this comment. The suggestion is in any case a compromise and it's probably best to wait until a proper solution exists before changing this formalization to be more universe polymorphic.

Reviewer: I reopened a discussion we had last week on Discord related this, because your formalization made me realize I had thought about this problem before. Maybe it elucidates the problem to you: https://discord.com/channels/954089080197611551/966602283351703562/1287029778645520415

Author: Should I formalize categories of short maps / isometries as _large_ categories? (are they?)

Reviewer: They are, but I don't know if it is worthwhile. See for instance what I write [here](https://github.com/UniMath/agda-unimath/blob/35e5915e3705e34ddb8f053196f2961853a08791/src/order-theory/precategory-of-posets.lagda.md#L31); your situation is completely analogous to posets. One could try to argue that it is more natural to consider metric structures using large categories as compared to your current approach, but I think it would still miss the mark.

Reviewer: I'll pre-emptively resolve this conversation.

>   is-equiv-is-iso-precatory-isometry-Metric-Space =
>     rec-Σ
>       ( λ g H →
>         is-equiv-is-invertible
>           ( map-isometry-Metric-Space B A g)
>           ( htpy-eq (ap (map-isometry-Metric-Space B B) (pr1 H)))
>           ( htpy-eq (ap (map-isometry-Metric-Space A A) (pr2 H))))

Reviewer: We prefer to define functions on Sigma types by pattern-matching

```suggestion
  is-equiv-is-iso-precatory-isometry-Metric-Space (g , H) =
    is-equiv-is-invertible
      ( map-isometry-Metric-Space B A g)
      ( htpy-eq (ap (map-isometry-Metric-Space B B) (pr1 H)))
      ( htpy-eq (ap (map-isometry-Metric-Space A A) (pr2 H)))
```

  equiv-iso-isometric-is-equiv-Metric-Space =
    equiv-tot (λ f → commutative-product) ∘e
    associative-Σ
      ( function-carrier-type-Metric-Space A B)
      ( is-isometry-Metric-Space A B)
      ( is-equiv ∘ map-isometry-Metric-Space A B) ∘e
    equiv-tot
      ( λ f →
        equiv-iff
          ( is-iso-prop-Precategory
            ( precategory-isometry-Metric-Space)
            { A}
            { B}
            ( f))
          ( is-equiv-Prop
            ( map-isometry-Metric-Space A B f))
          ( is-equiv-is-iso-precatory-isometry-Metric-Space A B f)
          ( is-iso-precategory-is-equiv-isometry-Metric-Space A B f))
```

### Isomorphism in the precategory of metric spaces and isometries is torsorial

```agda
module _
  {l1 l2 : Level} (A : Metric-Space l1 l2)
  where

  is-torsorial-iso-precategory-isometry-Metric-Space :
    is-torsorial (iso-Precategory precategory-isometry-Metric-Space A)
  is-torsorial-iso-precategory-isometry-Metric-Space =
    is-contr-equiv
      ( Σ (Metric-Space l1 l2) (isometric-is-equiv-Metric-Space A))
      ( equiv-tot (equiv-iso-isometric-is-equiv-Metric-Space A))
      ( is-torsorial-isometric-is-equiv-Metric-Space A)

> ```

Author: See 24d4350f07612235bac477a1efd9b3001082a7dc
I'm closing this and leaving the other conversation open, to discuss if we merge the (pre)categories modules, or not.

open import foundation.action-on-identifications-functions
open import foundation.contractible-types
open import foundation.dependent-pair-types
open import foundation.equivalences
open import foundation.function-extensionality
open import foundation.function-types
open import foundation.functoriality-dependent-pair-types
open import foundation.identity-types
open import foundation.logical-equivalences
open import foundation.propositions
open import foundation.subtypes
open import foundation.torsorial-type-families
open import foundation.transport-along-identifications
open import foundation.type-arithmetic-cartesian-product-types
open import foundation.type-arithmetic-dependent-pair-types
open import foundation.universe-levels

open import metric-spaces.equality-of-metric-spaces
open import metric-spaces.functions-metric-spaces
open import metric-spaces.isometry-metric-spaces
open import metric-spaces.metric-spaces
```

</details>

## Idea

The [isometries](metric-spaces.isometry-metric-spaces.md) form a
[precategory](category-theory.precategories.md).

## Definitions

```agda
module _
  {l1 l2 : Level}
  where

  precategory-isometry-Metric-Space :

>     Precategory (lsuc l1 ⊔ lsuc l2) (l1 ⊔ l2)

Reviewer: A nicer and easier way to formalize this category is probably to use wide subcategories.

Author: ok. got any reference module on that?

Reviewer: It seems wide subcategories aren't that much in use yet. Here's a reference module where the subcategory is full instead of wide, which gives an analogous story:
- [`ring-theory.category-of-cyclic-rings`](https://unimath.github.io/agda-unimath/ring-theory.category-of-cyclic-rings.html)

And here's a module that actually uses wide subcategories:
- [`category-theory.cores-categories`](https://unimath.github.io/agda-unimath/category-theory.cores-categories.html)

Reviewer: On second thought, since you don't define isometries as short maps with extra properties, it's probably best to leave the formalization as is. Wide subcategories are meant to be used when a class of maps is defined as a subclass of another class of maps
